torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1098 @@
+          case AttributeKind::gs:
+          case AttributeKind::ty:
+          case AttributeKind::tys:
+          case AttributeKind::c:
+          default:
+            TORCH_INTERNAL_ASSERT(
+                false,
+                "Unexpected attribute type ",
+                static_cast<int>(a->kindOf(attr)),
+                " from node ",
+                *a);
+            break;
+        }
+      };
+
+  for (const auto& it : func_ctx.attribute_map_) {
+    auto* ref_n = it.first;
+    for (const auto attr_it : it.second) {
+      const auto& attr = attr_it.first;
+      auto attr_name = func_ctx.FindAttrName(ref_n, attr).value();
+      copy_attr(ref_n, func_n, attr, attr_name);
+      for (auto* n : scope_ctx.nlist_) {
+        if (attr_it.second.find(n) != attr_it.second.end()) {
+          copy_attr(n, func_n, attr, attr_name);
+          break;
+        }
+      }
+    }
+  }
+
+  func_n->insertAfter(last_n);
+
+  return func_n;
+}
+
+void FunctionExtractor::ConvertScopeToFunction(
+    const ScopePtr& scope_key,
+    const scope_list& scope_list,
+    scope_ctx_map& scope_ctxs,
+    const std::shared_ptr<Graph>& graph) {
+  // This function needs to be called always on inner most scopes.
+  // 1. Generate function context, this identifies different constants and
+  // attributes.
+  // 2. Create function definition node, and insert to main graph.
+  // 3. Create function node for each call, and replace subgraph nodes in parent
+  // functions.
+
+  func_ctxs_.insert(std::make_pair(
+      scope_key, new FunctionContext(scope_key, scope_list, scope_ctxs)));
+  auto& func_ctx = *func_ctxs_[scope_key];
+
+  const std::string module_class_name(
+      func_ctx.scope_key_->name().toUnqualString());
+  auto pos = module_class_name.rfind(".");
+  TORCH_INTERNAL_ASSERT(pos != std::string::npos);
+
+  auto construct_unique_module_name = [&](std::string module_name) {
+    auto module_name_variant = module_variant_count_.find(module_name);
+    if (module_name_variant != module_variant_count_.end()) {
+      module_name += ("." + std::to_string(module_name_variant->second));
+    } else {
+      module_variant_count_[module_name] = 0;
+    }
+    module_variant_count_[module_name]++;
+    return module_name;
+  };
+
+  const auto domain_name = module_class_name.substr(0, pos);
+  const auto func_name =
+      construct_unique_module_name(module_class_name.substr(pos + 1));
+
+  auto func_def_n =
+      CreateFunctionDefNode(func_ctx, graph, domain_name, func_name);
+
+  // create and insert local function node to graph.
+  for (const auto& it : func_ctx.scope_ctxs_) {
+    auto scope = it.first;
+    auto& scope_ctx = *it.second;
+    auto func_n =
+        CreateFunctionNode(func_ctx, scope_ctx, graph, domain_name, func_name);
+
+    std::unordered_set<Node*> old_nodes(
+        scope_ctx.nlist_.begin(), scope_ctx.nlist_.end());
+
+    auto last_n = *scope_ctx.nlist_.rbegin();
+    // replace local function node in parent scopes.
```

Comment:
IIUC, this comment should read something like `// Replace function body in parent scopes with function call`.
My point is that `Replace X` normally means `Remove X and add Y`, but here the local function node is the thing being added (`Y`).